compression_test.go

```diff
@@ -0,0 +1,163 @@
+)
+
+type BiasedRandomReader struct {
+	sourceSize int
+	bytesRead  int
+}
+
+func NewBiasedRandomReader(sourceSize int) *BiasedRandomReader {
+	return &BiasedRandomReader{sourceSize, 0}
+}
+
+func (reader *BiasedRandomReader) Read(p []byte) (n int, err error) {
+	toRead := min(len(p), reader.sourceSize-reader.bytesRead)
+	for i := 0; i < toRead; i++ {
+		p[i] = byte(min(10, rand.Int()%256))
+	}
+	reader.bytesRead += toRead
+	if toRead < len(p) {
+		return toRead, io.EOF
+	}
+	return toRead, nil
+}
+
+type DifferentFileError struct {
+	firstFileContent  []byte
+	secondFileContent []byte
+}
+
+func (err DifferentFileError) Error() string {
+	return fmt.Sprintf("Files are different, but should be same, first file content:\n%v\nsecond file content:\n%v\n",
+		err.firstFileContent, err.secondFileContent)
+}
+
+func compressInitialFile(compressor Compressor, initialFilePath string, t *testing.T) error {
+	srcFile, err := os.Open(initialFilePath)
+	defer srcFile.Close()
+	if err != nil {
+		return err
+	}
+	compressedFile, err := os.Create(CompressedFilePath + compressor.FileExtension())
+	defer compressedFile.Close()
+	if err != nil {
+		return err
+	}
+	compressingWriter := compressor.NewWriter(compressedFile)
+	defer compressingWriter.Close()
+	bytesRead, err := compressingWriter.ReadFrom(srcFile)
+	t.Logf("Bytes read: %v", bytesRead)
+	return err
+}
+
+func decompressCompressedFile(compressorFileExtension string) error {
+	compressedFile, err := os.Open(CompressedFilePath + compressorFileExtension)
+	defer compressedFile.Close()
+	if err != nil {
+		return err
+	}
+	decompressedFile, err := os.Create(DecompressedFilePath)
+	defer decompressedFile.Close()
+	if err != nil {
+		return err
+	}
+	decompressor := FindDecompressor(compressorFileExtension)
+	err = decompressor.Decompress(decompressedFile, compressedFile)
+	return err
+}
+
+func compareDecompressedFileWithInitial(initialFilePath string) error {
+	decompressedFileContent, err := ioutil.ReadFile(DecompressedFilePath)
+	if err != nil {
+		return err
+	}
+	initialFileContent, err := ioutil.ReadFile(initialFilePath)
+	if err != nil {
+		return err
+	}
+	if bytes.Compare(decompressedFileContent, initialFileContent) != 0 {
+		return DifferentFileError{
+			firstFileContent:  decompressedFileContent,
+			secondFileContent: initialFileContent,
+		}
+	}
+	return nil
+}
+
+func testCompressor(compressor Compressor, initialFilePath string, t *testing.T) {
+
+	err := compressInitialFile(compressor, initialFilePath, t)
+	defer os.Remove(CompressedFilePath + compressor.FileExtension())
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	err = decompressCompressedFile(compressor.FileExtension())
+	defer os.Remove(DecompressedFilePath)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	err = compareDecompressedFileWithInitial(initialFilePath)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+}
+
+func TestSmallFileCompression(t *testing.T) {
+	for _, compressingAlgorithm := range compressingAlgorithms {
```

Comment:
How about adding lzop here too? If it is no a lot of work, of course. I know that it is somewhat broken #35 butI still hope to find an example of broken file.... in this case we should leave broken file even in an event of failure.